vendor/k8s.io/kubernetes/pkg/kubectl/resource_filter.go

```diff
@@ -0,0 +1,100 @@
+package kubectl
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type filterEntry struct {
+	filterFunc reflect.Value
+}
+
+// FilterOptions implements PrintOptions
+type FilterOptions struct {
+	PrintOptions
+}
+
+type ResourceFilter struct {
+	filterMap map[reflect.Type]*filterEntry
+	options   FilterOptions
+}
+
+func NewResourceFilter(opts *FilterOptions) *ResourceFilter {
+	filter := &ResourceFilter{
+		filterMap: make(map[reflect.Type]*filterEntry),
+		options:   *opts,
+	}
+
+	filter.addDefaultHandlers()
+	return filter
+}
+
+func (f *ResourceFilter) addDefaultHandlers() {
+	f.Handler(filterPod)
+}
+
+// Handler adds a filter handler to a ResourceFilter instance.
+// See validateFilterFunc for required method signature.
+func (f *ResourceFilter) Handler(filterFunc interface{}) error {
+	filterFuncValue := reflect.ValueOf(filterFunc)
+	if err := f.validateFilterFunc(filterFuncValue); err != nil {
+		glog.Errorf("Unable to add print handler: %v", err)
+		return err
+	}
+	objType := filterFuncValue.Type().In(0)
+	f.filterMap[objType] = &filterEntry{
+		filterFunc: filterFuncValue,
+	}
+	return nil
+}
+
+// validateFilterFunc validates the filter handler signature.
+// filterFunc is the function that will be called to filter an object.
+// It must be of the following type:
+//  func filterFunc(object ObjectType, options FilterOptions) error
+// where ObjectType is the type of the object that will be printed.
+func (f *ResourceFilter) validateFilterFunc(filterFunc reflect.Value) error {
```

Comment:
+1 to NOT using reflection here. Avoid coding for an uncertain future, keep it simple.

Checking if an arbitrary runtime value is a function, takes 2 arguments, etc, is ugly and inefficient.